presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java

```diff
@@ -40,16 +50,101 @@ private IpPrefixFunctions() {}
     @ScalarFunction("ip_prefix")
     @SqlType(StandardTypes.IPPREFIX)
     public static Slice ipPrefix(@SqlType(StandardTypes.IPADDRESS) Slice value, @SqlType(StandardTypes.BIGINT) long subnetSize)
+    {
+        InetAddress address = validateIpAddress(value);
+        validateSubnetSize(address.getAddress(), subnetSize);
+
+        return castFromVarcharToIpPrefix(utf8Slice(InetAddresses.toAddrString(address) + "/" + subnetSize));
+    }
+
+    @Description("IP prefix for a given IP address and subnet size")
+    @ScalarFunction("ip_prefix")
+    @LiteralParameters("x")
+    @SqlType(StandardTypes.IPPREFIX)
+    public static Slice stringIpPrefix(@SqlType("varchar(x)") Slice slice, @SqlType(StandardTypes.BIGINT) long subnetSize)
+    {
+        return ipPrefix(castFromVarcharToIpAddress(slice), subnetSize);
+    }
+
+    @Description("First/Smallest subnet IP address for a given IP prefix")
+    @ScalarFunction("ip_subnet_min")
+    @SqlType(StandardTypes.IPADDRESS)
+    public static Slice ipSubnetMin(@SqlType(StandardTypes.IPPREFIX) Slice value)
+    {
+        validateIpPrefixAndGetBytes(value);
+        return castFromIpPrefixToIpAddress(value);
+    }
+
+    @Description("Last/Maximum subnet IP address for a given IP prefix")
+    @ScalarFunction("ip_subnet_max")
+    @SqlType(StandardTypes.IPADDRESS)
+    public static Slice ipSubnetMax(@SqlType(StandardTypes.IPPREFIX) Slice value)
+    {
+        byte[] address = validateIpPrefixAndGetBytes(value);
+        int subnetSize = value.getByte(IPPREFIX.getFixedSize() - 1) & 0xFF;
+
+        if (address.length == 4) {
+            for (int i = 0; i < 4; i++) {
+                address[3 - i] |= (byte) ~(~0 << min(max((32 - subnetSize) - 8 * i, 0), 8));
+            }
+            byte[] bytes = new byte[16];
+            bytes[10] = (byte) 0xff;
+            bytes[11] = (byte) 0xff;
+            arraycopy(address, 0, bytes, 12, 4);
+            address = bytes;
+        }
+        else if (address.length == 16) {
+            for (int i = 0; i < 16; i++) {
+                address[15 - i] |= (byte) ~(~0 << min(max((128 - subnetSize) - 8 * i, 0), 8));
+            }
+        }
+        return wrappedBuffer(address);
+    }
+
+    @Description("Range of IP address's in the subnet of the IP prefix")
+    @ScalarFunction("ip_subnet_range")
+    @SqlType("array(IPADDRESS)")
+    public static Block ipSubnetRange(@SqlType(StandardTypes.IPPREFIX) Slice value)
+    {
+        validateIpPrefixAndGetBytes(value);
```

Comment:
The `value` is validated 3 times in this function.

Reply:
Removed the validation part. Since the IPREFIX input will always be avalid one.